src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/markdown/MarkdownProjectReader.kt

```diff
@@ -1,13 +1,110 @@
-        TODO("The MarkdownProjectReader isn't yet implemented.")
+    override fun constructProjectTree(container: ResourceContainer, project: Project)
+            : Pair<ImportResult, Tree> {
+        val projectRoot = container.dir.resolve(project.path)
+        return projectRoot
+                .buildFileTree()
+                .filterMarkdownFiles()
+                ?.map<Any> { f -> contentList(f) ?: collection(f, projectRoot) }
+                ?.flattenContent()
+                ?.let { Pair(ImportResult.SUCCESS, it) }
+                ?: Pair(ImportResult.LOAD_RC_ERROR, Tree(Unit))
+    }
+
+    private fun fileToId(f: File): Int =
+        f.nameWithoutExtension.toIntOrNull() ?: 0
+
+    private fun fileToSlug(f: File, root: File): String =
+        f.toRelativeString(root.parentFile)
+                .split('/', '\\')
+                .map { it.toIntOrNull()?.toString() ?: it }
+                .joinToString("_")
+
+    private fun bufferedReaderProvider(f: File): (() -> BufferedReader)? =
+            if (f.isFile) { { f.bufferedReader() } } else null
+
+    private fun collection(f: File, projectRoot: File): Collection =
+            collection(fileToSlug(f, projectRoot), fileToId(f))
+
+    private fun collection(slug: String, id: Int): Collection =
+            Collection(
+                    sort = id,
+                    slug = slug,
+                    labelKey = "helpcollection",
+                    titleKey = "$id",
+                    resourceContainer = null)
+
+    private fun content(sort: Int, label: String, id: Int, text: String): Content? =
+            if (text.isEmpty()) null
+            else Content(sort, label, id, id, null, text, FORMAT)
+
+    private fun contentList(f: File): List<Content>? =
+            bufferedReaderProvider(f)
+                    ?.let { contentList(it, fileToId(f)) }
+
+    private fun contentList(brp: () -> BufferedReader, fileId: Int): List<Content> {
+        val helpResources = brp().use { ParseMd.parse(it) }
+        var sort = 1
+        return helpResources.flatMap { helpResource ->
+            listOfNotNull(
+                    content(sort++, "title", fileId, helpResource.title),
+                    content(sort++, "body", fileId, helpResource.body)
+            )
+        }
     }
 }
+
+internal fun File.buildFileTree(): OtterTree<File> {
+    var treeRoot: OtterTree<File>? = null
+    val treeCursor = ArrayDeque<OtterTree<File>>()
+    this.walkTopDown()
+            .onEnter { newDir ->
+                OtterTree(newDir).let { newDirNode ->
+                    treeCursor.peek()?.addChild(newDirNode)
+                    treeCursor.push(newDirNode)
+                    true
+                }
+            }
+            .onLeave { treeRoot = treeCursor.pop() }
+            .filter { it.isFile }
+            .map { OtterTreeNode(it) }
+            .forEach { treeCursor.peek()?.addChild(it) }
```

Comment:
This solution is very intriguing to me! Certainly a different way of thinking than typical sequential operations.

Reply:
Yeah, I'm not super proud of my code readability here. It could be a bit cleaner if our OtterTree nodes had parent links, but my treeCursor stack serves the purpose.